core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -141,6 +145,93 @@ public RoutingNodes(ClusterState clusterState, boolean readOnly) {
         }
     }
 
+    private void addRecovery(ShardRouting routing) {
+        addRecovery(routing, 1, false);
+    }
+
+    private void removeRecovery(ShardRouting routing) {
+        addRecovery(routing, -1, false);
+    }
+
+    public void addInitialRecovery(ShardRouting routing) {
+        addRecovery(routing, 1, true);
+    }
+
+    private void addRecovery(ShardRouting routing, int howMany, boolean initializing) {
+        assert routing.initializing() : "routing must be initializing: " + routing;
+        assert howMany == -1 || howMany == 1 : howMany + " != math.abs(1)";
```

Comment:
If `howMany` can only be -1 or 1, why not make it a boolean increment/decrement flag?